From: Juergen Beisert <j.beisert@pengutronix.de>
Date: Tue, 9 Sep 2008 09:33:14 +0000 (+0200)
Subject: pcm038: Add ALSA audio support
X-Git-Url: http://pasiphae.extern.pengutronix.de/?p=imx%2Flinux-2.6.git;a=commitdiff_plain;h=bace6945efcbd449c22c8e85f5b375391de260b0

pcm038: Add ALSA audio support

Add the ALSA audio support to the BSP.

Signed-off-by: Juergen Beisert <j.beisert@pengutronix.de>
Signed-off-by: Sascha Hauer <s.hauer@pengutronix.de>
---

Index: linux-2.6.29/arch/arm/mach-mx2/pcm038.c
===================================================================
--- linux-2.6.29.orig/arch/arm/mach-mx2/pcm038.c	2009-03-31 10:00:11.000000000 +0200
+++ linux-2.6.29/arch/arm/mach-mx2/pcm038.c	2009-03-31 10:04:06.000000000 +0200
@@ -26,6 +26,7 @@
 #include <linux/delay.h>
 #include <linux/err.h>
 #include <linux/clk.h>
+#include <linux/dma-mapping.h>
 #include <asm/mach/arch.h>
 #include <asm/mach-types.h>
 #include <asm/mach/time.h>
@@ -39,6 +40,8 @@
 #include <mach/imx_ssi.h>
 #include <mach/pmic/platform.h>
 #include <mach/pmic/power.h>
+#include <mach/imx_dam.h>
+#include <mach/imx_sound.h>
 #include "cpufreq_imx.h"
 #include <mach/board-pcm038.h>
 #include <mach/mxc_nand.h>
@@ -484,9 +487,76 @@
 # define CPU_FREQUENCY
 #endif
 
+#ifdef CONFIG_SND
+
+/* ------------ sound is complicated ----------------- */
+
+static int mxc_ssi_pins[] = {
+	PC20_PF_SSI1_FS,
+	PC21_PF_SSI1_RXD,
+	PC22_PF_SSI1_TXD,
+	PC23_PF_SSI1_CLK,
+
+	PC16_PF_SSI4_FS,
+	PC17_PF_SSI4_RXD,
+	PC18_PF_SSI4_TXD,
+	PC19_PF_SSI4_CLK
+};
+
+static int gpio_ssi_activate(struct platform_device *pdev)
+{
+	int err;
+
+	err = mxc_gpio_setup_multiple_pins(mxc_ssi_pins, ARRAY_SIZE(mxc_ssi_pins),
+				MXC_GPIO_ALLOC_MODE_NORMAL, "imx-sound");
+	if (err < 0)
+		pr_err("Cannot register SSI pins for sound usage\n");
+
+	return err;
+}
+
+static struct imx_sound_platform_data pcm038_alsa_sound = {
+	.connection = {
+		[0] = {
+			.cpu_port = 1,	/* SSI1_* pin group */
+			.dev_port = 1	/* connected to port 1 of the PMIC */
+		},
+		[1] = {
+			.cpu_port = 4,	/* SSI4_* pin group */
+			.dev_port = 2	/* connected to port 2 of the PMIC */
+		}
+	},
+	.init = gpio_ssi_activate,
+	.exit = NULL	/* FIXME */
+};
+
+static struct platform_device pcm038_alsa_sound_device = {
+	.name = "imx-alsa",
+	.id = 0,
+	.dev = {
+		.platform_data = &pcm038_alsa_sound,
+		.coherent_dma_mask = DMA_BIT_MASK(32)
+	}
+};
+
+static struct imx_ssi_platform_data pcm038_ssi0 = {
+	.init = NULL,	/* FIXME currently done in the audio device */
+	.exit = NULL	/* FIXME */
+};
+
+static struct imx_ssi_platform_data pcm038_ssi1 = {
+	.init = NULL,	/* FIXME currently done in the audio device */
+	.exit = NULL	/* FIXME */
+};
+# define ALSA_SOUND &pcm038_alsa_sound_device,
+#else
+# define ALSA_SOUND
+#endif
+
 static struct platform_device *platform_devices[] __initdata = {
 	&pcm038_nor_mtd_device,
 	CPU_FREQUENCY
+	ALSA_SOUND
 	&mxc_w1_master_device,
 	&pcm038_sram_mtd_device,
 };
@@ -527,6 +597,12 @@
 				ARRAY_SIZE(pcm038_spi_board_info));
 #endif
 
+#ifdef CONFIG_SND
+	mxc_register_device(&mxc_dam_device, NULL);
+	mxc_register_device(&imx_ssi_device0, &pcm038_ssi0);
+	mxc_register_device(&imx_ssi_device1, &pcm038_ssi1);
+#endif
+
 	platform_add_devices(platform_devices, ARRAY_SIZE(platform_devices));
 
 #ifdef CONFIG_MACH_PCM970_BASEBOARD
